### ISSUE TITLE
Multiple disjoint meshes do not work (occlusion fails)

### ISSUE THREAD

**Issue by ebartrum** [nv-tlabs/DIB-R#9]
I'm currently rendering using RenderMeshColor as the rendering class. If I render 2 meshes, the nearer mesh does not occlude the further away mesh. I realise now that this is causing the issue in https://github.com/nv-tlabs/DIB-R/issues/8

Here is an example:
![rotation_test2](https://user-images.githubusercontent.com/9696427/69892221-3431ed80-12fb-11ea-8273-8ffa86219b88.gif)

How can I render multiple meshes with vertex colour, with the correct occlusion?
Thanks!

**Comment by wenzhengchen** [comment]
For now, we only support 1 mesh. When you render 2 meshes, how do you do it? Do you convert it into one mesh, or draw it separately? IF it is later, then it would have occlusion problem.

**Comment by ebartrum** [comment]
I convert the 2 meshes into one mesh; ie combine the separate lists of vertices and faces and textures into a single list of vertices, faces and textures. I have looked through the code and I don't see why it doesn't work in this multiple mesh case.

**Comment by wenzhengchen** [comment]
So instead of drawing the nearest faces, it draws farthest faces?
You may check here:
https://github.com/nv-tlabs/DIB-R/blob/611f69dce00e75f65b37722120010e2da46fc1bf/dib-render/cuda_dib_render/dr_cuda_for.cu#L142

In this line, we check the face depth to just draw the nearest faces. I am not sure what happens to your mesh. Can you also visualize it in mesh lab to double-check? Our render should be the same as  meshlab.

**Comment by ebartrum** [comment]
Here is a script to reproduce the error. I create 2 spheres, one red and one blue. I position the camera so that the red should occlude the blue. I render it and the blue sphere is visible, even though it should be occluded.

`
from DIB_R.utils.render.renderfunc_cluster import RenderMeshColor
from DIB_R.dib_render.render_cuda.utils_render_color2 import linear
import meshzoo
import torch
import matplotlib.pyplot as plt

# Sphere1 has radius 1, centre at (0,0,0). Colour is red
vertices1, faces1 = meshzoo.uv_sphere(num_points_per_circle=20, num_circles=10, radius=1.0)
vertices1 = torch.from_numpy(vertices1).float().cuda()
vertices1 = vertices1.unsqueeze(0)
faces1 = torch.from_numpy(faces1).long().cuda()
texture1 = torch.ones(vertices1.shape).cuda() * torch.tensor([1.,0,0]).cuda()

# Sphere2 has radius 1, centre at (0,0,5). Colour is blue
vertices2 = vertices1 + torch.tensor([0,0,5.]).cuda()
faces2 = faces1 + faces1.max() + 1
texture2 = torch.ones(vertices2.shape).cuda() * torch.tensor([0,0,1.]).cuda()

#Combine the meshes
vertices = torch.cat([vertices1, vertices2], dim=1)
faces = torch.cat([faces1, faces2], dim=0)
texture = torch.cat([texture1, texture2], dim=1)

# Position camera at (0,0,-2). Sphere1 should occlude sphere2.
camera_rot = torch.eye(3).unsqueeze(0).cuda()
camera_pos = torch.tensor([0,0,-2]).unsqueeze(0).cuda().float()
focal_length = 1.5
camera_proj = torch.tensor([1,1,1/focal_length]).unsqueeze(0).cuda()
renderer = RenderMeshColor(linear)

rendered = renderer(vertices, texture,
        faces, camera_rot, camera_pos, camera_proj)
rendered_image = rendered[0][0][0] 

# In rendered image, sphere2 is not occluded.
plt.imshow(rendered_image.cpu())
plt.show()
`

Here is the output that I see:
![reproduce_error](https://user-images.githubusercontent.com/9696427/69922284-f8319080-1492-11ea-812d-5a9268615481.png)

**Comment by ebartrum** [comment]
I realised that the issue above is due to the camera being on the wrong side, and the objects being behind the camera. It seems then that the issue is: when objects are behind the camera, they are shown, with occlusion occurring in the reverse order.
